### PR TITLE
Ocm pullsecret validation

### PR DESCRIPTION
**Pull request by facchettos**
### Which issue this PR addresses:
https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/13933226/

### What this PR does / why we need it:

Creates a new validating webhook on the clusters to validate that pullsecrets are valid and rejects them if they are not. 
It creates it own namespace, uses service certificates and register the hook on the kubernetes API. 
It is possible to ignore registries that are known to be down (one of the ones from the OCM pullsecret is down for example). When the pullsecret is received, the hook will check that we actually get a successful auth on the registry, but it does not check anything else. In some cases , the pullsecret could be valid but not for some images, but unless we know all the images that are to be accessed, we can not do anything more. 

This also prevents modification of the aro registry field in the pull secret (same goal as the reconciler, but instead of recovering it, prevents the modification). It can run alongside the reconciler for now without interfering, and we will be able to remove the reconciler later. 

### Test plan for issue:
Wrote unit tests, and tested on a cluster. 

### Is there any documentation that needs to be updated for this PR?
Once we are ready to merge, I'll update the docs

### CODE REVIEW

**Review by s-amann — CHANGES_REQUESTED** [review]
A few questions, otherwise lgtm

File: pkg/operator/admission/validation/pullsecret/validate_test.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/validate_test.go:149]: actually, it should be 401. My bad, fixing it

**Comment by facchettos** [comment]
golangci-lint is failing but I think it is the action itself that has problems, because running it locally does not yield any problem

**Review by s-amann — DISMISSED** [review]

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (jmilhau) [inline pkg/operator/admission/validation/pullsecret/secret_utils.go:45]: OCM

**Review by jhoreman — DISMISSED** [review]
LGTM, thanks for this.

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:45]: I'll fix it once we get approval from leads, don't want to dismiss reviews from other engineers yet

File: pkg/operator/admission/validation/pullsecret/validator_test.go
Comment (AldoFusterTurpin) [inline pkg/operator/admission/validation/pullsecret/validator_test.go:116]: Same as before 🙂

File: pkg/operator/admission/validation/pullsecret/validator_test.go
Comment (AldoFusterTurpin) [inline pkg/operator/admission/validation/pullsecret/validator_test.go:165]: I also like to split that into multiple lines. Nice👌

File: pkg/operator/admission/validation/pullsecret/validator_test.go
Comment (AldoFusterTurpin) [inline pkg/operator/admission/validation/pullsecret/validator_test.go:240]: We could use an "or" to simplify it a bit.

**Review by AldoFusterTurpin — COMMENTED** [review]
@facchettos Great work 👏 🚀. 

I have added some suggestions: some are just minor changes, some are little improvements and one important concern about concurrency ("validateCredentials" function).

Thanks! 🙂

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/http_server.go:43]: this is a setup function, there isn't really a need for that.

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/http_server.go:56]: if we have it as a package level it become accessible to other things too. and we can't have const maps.

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/http_server.go:60]: this is not useful.

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/http_server.go:111]: this makes it more complicated.

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:90]: base64 has a very limited set of characters(64) , which are all ascii characters.

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:42]: this kind of functions is confusing. it's better to have explicit returns

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:105]: same answers

File: pkg/operator/admission/validation/pullsecret/validator.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/validator.go:193]: `routineResult := <-results` is actually a common pattern. 

`for range` does not prevent any blocking. because it will wait for the channel to be closed it will block and will be a deadlock because we haven't closed the channel at that point. 

I think we need all the errors, at least for logging to see all repositories that are failing, will add that 

Your suggestion with the waitgroup is equivalent to the current one, just with an extra step.

File: pkg/operator/admission/validation/pullsecret/validator_test.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/validator_test.go:116]: same as before :slightly_smiling_face:

File: cmd/aro/operator.go
Comment (facchettos) [reply cmd/aro/operator.go:232]: consistency with the previous statements

File: pkg/operator/admission/validation/pullsecret/validator_test.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/validator_test.go:240]: same as the others, when it fails it will say at what line, and make it easier to know what part of the test failed

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:89]: will fix

File: pkg/operator/admission/validation/pullsecret/validator.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/validator.go:159]: will fix

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:73]: these are used exactly once, so there isn't really a point in doing it.

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:42]: also, as per https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/ , `# operation can be CREATE, UPDATE, DELETE, or CONNECT`

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/http_server.go:54]: will make it a const

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (AldoFusterTurpin) [reply pkg/operator/admission/validation/pullsecret/http_server.go:44]: Yes, I understand. There is not a need for that but then your function will not work if the image is not an env variable. I still think the domain logic should not depend on that detail. But just a subtle improvement. Np

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (AldoFusterTurpin) [reply pkg/operator/admission/validation/pullsecret/http_server.go:57]: Yes, I know we can't have constant maps in Go.  

Not saying that should be done but be aware that if you ever need some kind of constant map in a more complicated case,  you can always use a regular function and call that function directly in StartValidator
```go
func getRequired() map[string]bool {
   map[string]bool{"quay.io": true}
}

func StartValidator(ctx context.Context, log *logrus.Entry, certPath, keyPath string) error {
	c := ociRegClient{
		httpClient: &http.Client{Timeout: httpTimeout},
		ctx:        ctx,
		required:  getRequired(),
		log:        log,
	}
	http.HandleFunc("/", c.handleRequest)
	return http.ListenAndServeTLS(":8080", certPath, keyPath, nil)
}
```
That way no one can modify the map value if you always invoke directly the function.

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (AldoFusterTurpin) [reply pkg/operator/admission/validation/pullsecret/http_server.go:118]: Okay. Why it is more complicated? 🙂
I don't understand why it is more complicated if you just have one return in the function without any if condition and without modifying an already instantiated struct.

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (AldoFusterTurpin) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:43]: I see, okay. I thought the two if's were exclusive, my fault. 

"this kind of functions is confusing. it's better to have explicit returns and variable assignments" -> I think the use of named return parameters still applies as you take advantage of the zeroed values of the variables that you want to return. If you want explicit returns just add them 🙂
```go

func unmarshalRequestToSecret(request *admissionv1.AdmissionRequest) (secret corev1.Secret, oldSecret corev1.Secret, err error) {
	if request.Operation != admissionv1.Delete {
		err = json.Unmarshal(request.Object.Raw, &secret)
		if err != nil {
			return secret, oldSecret, err
		}
	}

	if request.Operation == admissionv1.Update || request.Operation == admissionv1.Delete {
		err = json.Unmarshal(request.OldObject.Raw, &oldSecret)
		if err != nil {
			return secret, oldSecret, err
		}
	}

	return secret, oldSecret, nil
}
```

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (AldoFusterTurpin) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go]: Right! Just wanted confirmation. Tks!

File: pkg/operator/admission/validation/pullsecret/validator.go
Comment (AldoFusterTurpin) [reply pkg/operator/admission/validation/pullsecret/validator.go:192]: "Also, your solution with an errgroup does not run in parallel because you have not set the limit. So it is not all that much simpler in the end." 

I think this statement is incorrect, [example](https://go.dev/play/p/d3o6Z420M0p) of the official [docs](https://pkg.go.dev/golang.org/x/sync/errgroup#example-Group-JustErrors) uses it without the setLimit. In fact, this is the [source code](https://cs.opensource.google/go/x/sync/+/master:errgroup/errgroup.go;l=60). 

comment of the source code -> "// SetLimit limits the number of active goroutines in this group to at most n."

so it's something that isn't needed here as you want all to run concurrently. 

"Your suggestion with the waitgroup is equivalent to the current one, just with an extra step, and that we now need to pass the waitgroup to the routines." 

No, you don't need to pass the waitGroup to the function, you enclose the call to the go routine in an anonymous function so it can access the waitGroup from inside the anonymous function, like [here](https://gobyexample.com/waitgroups) (this example is just to show we don't need to pass the waitGroup, I know this example is a different case).

"for range does not prevent any blocking. in fact it will wait for the channel to be closed it will block and will be a deadlock because we haven't closed the channel at that point."

Obviously using a "for range" with your code will have a deadlock as you close the channel from the receiver part (something that in general is not a good practice), I didn't propose that. I suggested using a waitgroup + "for range" as it is more idiomatic and clear.

If you want all the errors, I would use a waitGroup like in the official [docs](https://pkg.go.dev/sync#example-WaitGroup) (please note that the wait group is accessible from the anonymous function as I mention before) + the errors channel that you are already using, but using a manual counter to wait on go routines is something that should be avoided as can lead to problems and is not idiomatic.

🙂

File: pkg/operator/admission/validation/pullsecret/validator_test.go
Comment (AldoFusterTurpin) [reply pkg/operator/admission/validation/pullsecret/validator_test.go:116]: Same as before 🙂 
(😂)

File: pkg/operator/admission/validation/pullsecret/validator.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/validator.go:192]: putting the waitgroup in an anonymous function is even worse. and does not bring *any* value to this code. We know how many results we get , we read as many as we send, it's that simple. why do you want to add a useless step ?

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/http_server.go:57]: I know how to do it, however we don't need it as a constant, we just feed it to that one struct. stop trying to make the code more complicated than it has to be.

File: pkg/operator/admission/validation/pullsecret/http_server.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/http_server.go:61]: this is just wrong Aldo, it makes the code more complicated to read because you have to go back and forth between different places in the code. You are just suggesting to make the code more complicated and less readable.

**Comment by hawkowl** [comment]
@facchettos I'm not sure why we're implementing this in a validating webhook when we have the Guardrails project for similar things? That got put on hold for now, and I'm guessing this would be too...

**Comment by jharrington22** [comment]
/hold

Putting a hold on this PR for the moment as it is duplicating what Guardrails was intended to solve. cc/ @rogbas

**Comment by facchettos** [comment]
guardrails uses gatekeeper. Gatekeeper is not a good fit for that use case.

**Comment by hawkowl** [comment]
@facchettos This would break using staging. OCM pullsecrets from staging are still `cloud.openshift.com`. You would need to figure out the Telemeter URL to properly check the secret.

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (hawkowl) [inline pkg/operator/admission/validation/pullsecret/secret_utils.go:122]: Why not? This is a valid action.

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (hawkowl) [inline pkg/operator/admission/validation/pullsecret/secret_utils.go:127]: this will need to be cloud aware (in case we use other registries)

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:122]: this is not ? this is the whole point of the work item to prevent deletion of the secret.

File: pkg/operator/admission/validation/pullsecret/secret_utils.go
Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go:127]: just so I understand, this should be env aware so for example if we use arosvcint in int, it does not fail ?

Comment (facchettos) [reply pkg/operator/admission/validation/pullsecret/secret_utils.go]: fixed

**Pull request closed without merging**
